src/frontend/src/optimizer/logical_optimization.rs

```diff
@@ -0,0 +1,545 @@
+        plan = plan.optimize_by_rules(&JOIN_REORDER);
+
+        // Predicate Push-down: apply filter pushdown rules again since we pullup all join
+        // conditions into a filter above the multijoin.
+        plan = plan.predicate_pushdown(
+            Condition::true_cond(),
+            &mut PredicatePushdownContext::new(plan.clone()),
+        );
+        if explain_trace {
+            ctx.trace("Predicate Push Down:");
+            ctx.trace(plan.explain_to_string().unwrap());
+        }
+
+        // For stream, push down predicates with now into a left-semi join
+        plan = plan.optimize_by_rules(&FILTER_WITH_NOW_TO_JOIN);
+
+        // Push down the calculation of inputs of join's condition.
+        plan = plan.optimize_by_rules(&PUSH_CALC_OF_JOIN);
+
+        // Prune Columns
+        let required_cols = (0..plan.schema().len()).collect_vec();
+        let mut column_pruning_ctx = ColumnPruningContext::new(plan.clone());
+        plan = plan.prune_col(&required_cols, &mut column_pruning_ctx);
+        // Column pruning may introduce additional projects, and filter can be pushed again.
+        if explain_trace {
+            ctx.trace("Prune Columns:");
+            ctx.trace(plan.explain_to_string().unwrap());
+        }
+
+        if column_pruning_ctx.need_second_round() {
+            // Second round of column pruning and reuse the column pruning context.
+            // Try to replace original share operator with the new one.
+            plan = plan.prune_col(&required_cols, &mut column_pruning_ctx);
+            if explain_trace {
+                ctx.trace("Prune Columns (For DAG):");
+                ctx.trace(plan.explain_to_string().unwrap());
+            }
+        }
+
+        plan = plan.predicate_pushdown(
+            Condition::true_cond(),
+            &mut PredicatePushdownContext::new(plan.clone()),
+        );
+        if explain_trace {
+            ctx.trace("Predicate Push Down:");
+            ctx.trace(plan.explain_to_string().unwrap());
+        }
+
+        // Convert distinct aggregates.
+        plan = plan.optimize_by_rules(&CONVERT_DISTINCT_AGG_FOR_STREAM);
+
+        plan = plan.optimize_by_rules(&JOIN_COMMUTE);
+
+        plan = plan.optimize_by_rules(&PROJECT_REMOVE);
+
+        plan = plan.optimize_by_rules(&CONVERT_WINDOW_AGG);
+
+        if has_logical_over_agg(plan.clone()) {
+            return Err(ErrorCode::InternalError(format!(
+                "OverAgg can not be transformed. Plan:\n{}",
+                plan.explain_to_string().unwrap()
+            ))
+            .into());
+        }
+
+        plan = plan.optimize_by_rules(&DEDUP_GROUP_KEYS);
+
+        #[cfg(debug_assertions)]
+        InputRefValidator.validate(plan.clone());
+
+        if ctx.is_explain_logical() {
+            ctx.store_logical(plan.explain_to_string().unwrap());
+        }
+
+        Ok(plan)
+    }
+
+    pub fn gen_optimized_logical_plan_for_batch(mut plan: PlanRef) -> Result<PlanRef> {
```

Comment:
Hmmm. So it comes to another problem - lots of duplication code between the `for_batch` and `for_stream`. I am not sure whether it's better than before.

Reply:
Somewhat agree. I think it's easier to understand in a single block and branching on batch or stream. It makes it clearer which optimizations are shared, and which apply only to batch or stream. 

But this assumes that the general ordering of optimization passes are shared (so far, it's been true, but I can't tell if this PR also changes that)

Reply:
> Somewhat agree. I think it's easier to understand in a single block and branching on batch or stream. It makes it clearer which optimizations are shared, and which apply respectively only to batch or stream.
> 
> But this assumes that the general ordering of optimization passes are shared (so far, it's been true, but I can't tell if this PR also changes that)

In the beginning, it sounds good to only have a single block and branching on batch or stream, however, It contains more and more branches gradually. At least for now, I can give about 5 optimizations that needed to be applied to the batch and stream separately. In addition, I hope this separation can help developers to think of whether their optimization can only be applied to the batch or the streaming, or both.